### PR TITLE
[Issue-#89] Fix call of main methods (refer to the actual method owner not just the method)

### LINKED ISSUE #89: Support main methods from traits in Scala 3.4 (200USD Bounty)
---

**From the maintainer Li Haoyi**: I'm putting a 200USD bounty on this issue, payable by bank transfer on a merged PR resolving this issue, including a unit test to avoid regression. The issue is straightforward to reproduce.

---

My Main object has grown too large and I want to split it. I tried to create a trait for each entry point and then making the Main object extends all those traits. But in such case, mainargs does not find any of the @main annotated entry points in any of the traits, while it worked flawlessly when all those methods were together in the same Main object.

```scala
trait CommandList {
  @main
  def list(@arg v: String): Unit = ???
}

trait CommandCopy {
  @main
  def copy(@arg from: String, @arg to: String): Unit = ???
}

object Main extends CommandList with CommandCopy {
  def main(args: Array[String]): Unit = ParserForMethods(this).runOrExit(args)
}
```

Is there any way to support this use case? Without having to duplicate the entry point in all of them, of course.

### PR DESCRIPTION
**Pull request by arainko**
Fixes #89

### CODE REVIEW

**Comment by arainko** [comment]
The culprit was exactly as described in https://github.com/com-lihaoyi/mainargs/issues/89#issuecomment-2245607912

**Comment by lihaoyi** [comment]
@arainko could you isolate the semantic change and skip the autoformatting? We can autoformat in a separate PR if we want, but as is it makes this PR very hard to review

File: mainargs/src-3/Macros.scala [hunk 1]
```diff
@@ -89,8 +89,11 @@ object Macros {
     val argSigs = Expr.ofList(argSigsExprs)
 
     val invokeRaw: Expr[(B, Seq[Any]) => T] = {
-      def callOf(args: Expr[Seq[Any]]) = call(method, '{ Seq( ${ args }) }).asExprOf[T]
-      '{ ((b: B, params: Seq[Any]) => ${ callOf('{ params }) }) }
+
+      def callOf(methodOwner: Expr[Any], args: Expr[Seq[Any]]) =
```
Comment (arainko) [inline mainargs/src-3/Macros.scala:93]: `methodOwner` is the actual instance of an object that owns the main methods to call

File: mainargs/src-3/Macros.scala [hunk 2]
```diff
@@ -137,12 +139,9 @@ object Macros {
       }
     }
 
-    val base = Apply(fct, accesses.head)
-    val application: Apply = accesses.tail.foldLeft(base)((lhs, args) => Apply(lhs, args))
-    val expr = application.asExpr
-    expr
+    methodOwner.asTerm.select(method).appliedToArgss(accesses).asExpr
```
Comment (arainko) [inline mainargs/src-3/Macros.scala:142]: we first select the method using its `Symbol` on the owner and apply all of its parameter lists using `.appliedToArgss` as opposed to handrolling this tree

**Comment by lihaoyi** [comment]
Thanks @arainko! Can send me your bank transfer details to haoyi.sg@gmail.com and I can close out the bounty

**Pull request merged**